### PR TITLE
Admin, Enterprise Settings, Social: Add instagram handler validator

### LINKED ISSUE #1760: Instagram links not working
## Description
In Enterprise Setting > Social there is a field to insert an Instagram link. This link will appear on enterprise profiles. If the instagram page is https://www.instagram.com/happyfarm/ the link needs to be in the format "happyfarm". 
image
Currently there is no pointer telling the user that this is the required format. As a results, there are multiple different ways that people include the link, most of which lead to broken links on their profile.

Link formats that will work: happyfarm
Link formats that do not work: /happyfarm, @happyfarm, https://www.instagram.com/happyfarm/ and www.instagram.com/happyfarm/

## Expected Behavior
Users can find out what format the link needs to be in.
If a user puts in a link in a format that's wrong, the link either corrects, or there is an error message.

## Actual Behavior
Users don't know what to enter. 
If they do enter an incorrect format there's no error message, so they don't know that the link will be broken. The link also doesn't autocorrect, so the instance can have many many profiles with broken links, which looks unprofessional.

## Steps to Reproduce
1. Go to any enterprise settings, and go to the "social" tab
2. See instagram field, no suggestion is shown.
3. Try to put in an incorrect format. You'll notice that the link is accepted even if wrong. Look the result from frontend on your profile, you'll see the link is broken.
4. Try to put Instagram link in the "happyfarm" only format, and look the result on the frontend. Link works.

## Context
This issue is leading to lots of broke links in the directory. 
It's also frustrating for users who are creating their profile because they don't know what format we want the link in.

## Severity
S4 (UX issue I would say as the software is working as expected but the user doesn't know what to do)

## Your Environment
PC, French production, Chrome

## Possible Fix
As at Noverber 9th 2018 the following fix has been decided. We tried to implement this fix via PRs https://github.com/openfoodfoundation/openfoodnetwork/pull/2859 and https://github.com/openfoodfoundation/openfoodnetwork/pull/2849, but as mentioned on 2849 there were some issues. 

Fix 1
- When a user first enters their social media links, if the instagram link is in any of the following four formats (@happyfarm, /happyfarm, https://www.instagram.com/happyfarm/ or www.instagram.com/happyfarm/) the form should convert the link to the correct shorthand format (happyfarm)
- If the system can't recognise the link, and can't convert it, there should be an error message when the user saves the form.
image

Fix 2 - 
- We should add a suggestion to the Instagram field so that the user knows what format to use.
image
- We should ensure that any pre-existing profiles that have links in any of the four formats (@happyfarm, /happyfarm, https://www.instagram.com/happyfarm/ or www.instagram.com/happyfarm/) are converted to the correct shorthand format (happyfarm)
- If a pre-existing profile has an instagram link in a format which cannot be converted, the next time that user tries to save their Enterprise settings, they should see the error message pictured above.

Note: Fix 1 and Fix 2 should be merged close to each other in time. If Fix 1 is merged without Fix 2, pre-existing profiles with incorrect links will get an error message without knowing what the correct format their link should be in.

### PR DESCRIPTION
**Pull request by Laurel16**
#### What? Why?

- Closes #1760 

I looked at the (long 🙂) history of this issue and resumed this closed PR https://github.com/openfoodfoundation/openfoodnetwork/pull/9075, trying to make some improvements:

1. Modifiying the validations functions in entreprise.rb to better match the pattern we want. 

  Now patterns like: @my_user, http://instagram.com/user, https://www.instagram.com/user, instagram.com/@user, Https://www.Instagram.com.   /@User, instagram.com/user are transmormed to "user" during the validation process. 

  Patterns like: https://facebook.com/user, tagram.com/user, https://instagram.com/user/preferences, https://instagram.com /user/ are marked as invalid. 

2- Modifying the migration file to use the model validations.

I kept the migration test file but it is a bit useless because the modifications on the instances are already done by the validations of the model at the time of their creation (I didn't find how to use invalid instances for the tests). 

**Test the migration**

Before deploy on staging (prefer uk staging since it has enterprises with instagram profile), run the following query:

`select instagram from enterprises where instagram <> ''` ; keep the results.

deploy (the migration should run) and then compare with the new results of the previous sql query. We should not spot any errors. 

#### What should we test?

- In Admin, go to any enterprise settings, and go to the "social" tab
- See instagram field, try to put "https://instagram.com/user/preferences". You should see an error message top of the page (in case of errors, you are redirect to the entreprise settings page. A TODO fix ?). 
- Now, try to put "instagram.com/@user" in the instagram field. The value should be accepted and and returning to the tab social, you should see "user" in the instagram field. 
    
 I hope I have understood the ins and outs of this. I'm looking forward to hearing from you if I haven't. Thanks!

### CODE REVIEW

**Comment by jibees** [comment]
Seems like rubocop linter complains, and specs are actually failing. Back to `In Dev` state then.

File: spec/migrations/update_enterprise_instagram_links.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+require Rails.root.join('db/migrate/20221101201147_update_entreprise_instagram_links.rb')
+
+describe UpdateEntrepriseInstagramLinks do
```
Comment (jibees) [inline spec/migrations/update_enterprise_instagram_links.rb]: this file should contains `_spec` at the end of its name (like it's done in the folder `spec/migrations`)

File: config/initializers/form_error.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+#
+# You can configure these options here.
+
+config = {
+  error: 'error',
+  invalid: 'invalid',
+  template: '<span class="{error}"><br>{message}</span>'
+}
+
+ActionView::Base.field_error_proc =
```
Comment (jibees) [inline config/initializers/form_error.rb:15]: I'm wondering why this was necessary. Don't we already have any error for form inputs?

**Review by jibees — CHANGES_REQUESTED** [review]
Thanks @Laurel16 !

Quick comments: 
 - I guess some commits needs to be squashed together to simplify code review and git history. It is quite hard to review since file changed some commit after, ... 
 - Don't think `Test: ` and `Fix: ` are useful in the commit message. Would vote for direct message.

Thanks a lot !

File: config/initializers/form_error.rb (hunk 2 above)
Comment (Laurel16) [reply config/initializers/form_error.rb:15]: Yes it's just a proposal to have error messages under the inputs but maybe too much in one branch. I rollback.

Comment (jibees) [reply config/initializers/form_error.rb:15]: Yeap I agree: should be another PR then. Thanks!

**Comment by Laurel16** [comment]
> Thanks! One more think ... sorry but I think those two commits
> 
> `Merge branch 'master' into instagram` and `rebase`
> 
> Should be squashed/removed/renamed don't know exactly... ;)

Done, thanks!

**Review by jibees — APPROVED** [review]

**Comment by drummer83** [comment]
Hi @Laurel16, so great to see someone is finally pushing this over the line! :wink: 

I left a comment in the code, maybe you can have a quick look and let me know what you are thinking before I stage this on our server and run the migration.

Thanks for your work on this! :green_heart:

**Comment by Laurel16** [comment]
> Hi @Laurel16, so great to see someone is finally pushing this over the line! 😉
> 
> I left a comment in the code, maybe you can have a quick look and let me know what you are thinking before I stage this on our server and run the migration.
> 
> Thanks for your work on this! 💚

Thank you @jibees, sorry I can't see you comment, can you please specify where it is?

**Comment by drummer83** [comment]
It was me, @Laurel16. 😉
https://github.com/openfoodfoundation/openfoodnetwork/pull/9933#discussion_r1031555924

**Comment by Laurel16** [comment]
> It was me, @Laurel16. 😉 [#9933 (comment)](https://github.com/openfoodfoundation/openfoodnetwork/pull/9933#discussion_r1031555924)
> 
> Second comment.

I'm really sorry @drummer83. I can partially see your thumbnail comment but I can't find it in the comments list 🤔

**Comment by sigmundpetersen** [comment]
I can't see your comment either @drummer83 
>https://github.com/openfoodfoundation/openfoodnetwork/pull/9933#discussion_r1031555924

does not work 🤷

**Comment by drummer83** [comment]
That's strange. Maybe it's related to the mobile app. Here is a screenshot:

**Comment by sigmundpetersen** [comment]
Guess it's due to that yellow 'Pending' sign above your message 🙂 🤷

**Review by drummer83 — COMMENTED** [review]
Maybe now you can see my comments?

**Comment by drummer83** [comment]
@Laurel16 I think I found it. Could you please have a look again above? Sorry for the hassle.

**Comment by Laurel16** [comment]
> @Laurel16 I think I found it. Could you please have a look again above? Sorry for the hassle. Thanks @sigmundpetersen for pointing me at the solution.

Thanks @drummer83 and to all. Good catch with the "-" character. I upadted the regex and add a test. 
The regex %r{[^a-z0-9_\.]{1,30}}i doesn't work for me and doesn't seems to catch the - pattern, does it?

File: app/models/enterprise.rb [hunk 3]
```diff
@@ -461,15 +462,15 @@ def send_welcome_email
   end
 
   def strip_url(url)
-    url&.sub(%r{(https?://)?}, '')
+    url&.sub(%r{(https?://)|(http?://)?}, '')
```
[the remark this answers is not in the document]
Comment (Laurel16) [reply app/models/enterprise.rb]: yes, I removed it 🙏

**Comment by drummer83** [comment]
Sorry for the late reply, @Laurel16!
You are totally right, sorry! My Regex above was checking for INVALID characters, that's why I had the negation. 👍 

I am moving this to a quick Code Review again due to your latest changes. Maybe you would like to check yourself first if it is intentional to have lines 19-30 doubled in lines 32-43 in app/models/enterprise.rb?

Thanks for your patience!

**Comment by Laurel16** [comment]
> 

No it wasn't 🙂 Removed, thanks!

**Comment by sigmundpetersen** [comment]
Sorry, my bad.

Now I see there is a conflict that must be resolved though.
I guess a rebase would do.

**Comment by RachL** [comment]
@jibees if you can, can you rebase + review? I would hate to postpone again this PR moving forward given how many contributors tried to get a PR through for this one. So the quicker we can test and merge, the better :)

**Review by jibees — APPROVED** [review]
Looks good to me. 
We should have a look at the migration, just to be sure that none of the instagram url will be _damaged_.

**Comment by filipefurtad0** [comment]
Hey @Laurel16 ,

Thanks for your work here :clap: :clap: :clap: 

Below the ran tests on:

**The DB**

I've fetched the query before and after running the migration - I've attached the this comment.

The ordering on the results seems to change between migrations - not related to the PR itself - so I've changed the query slightly to include the enterprise name, to make sure we have the correct results.

[instagram_db_query_before_migration.csv](https://github.com/openfoodfoundation/openfoodnetwork/files/10268230/instagram_db_query_before_migration.csv)
[instagram_db_query_after_migration.csv](https://github.com/openfoodfoundation/openfoodnetwork/files/10268231/instagram_db_query_after_migration.csv)

The migration seems to work, and fetch the instagram handle correctly in most cases. I've found that entries starting with `www.instagram/` remain unchanged, see for example:

`www.instagram/thedenjuiceco`
`www.instagram/fernstone_naturals`
...

But I guess it's impossible to fetch the handle in all incorrect cases. In any case, the PR leaves these entries unchanged.

Most importantly: I could not spot any incorrect data migration.

**The UI**

> Now patterns like: @my_user, http://instagram.com/user, https://www.instagram.com/user, instagram.com/@user, https://www.instagram.com/. /@user, instagram.com/user are transformed to "user" during the validation process.

This is well covered on your model spec - which is great. I've had a quick look to see how does this handle on the UI, so typing the strings below and hitting "Update" transform the string or, where indicated, triggers the error message:

`@my_user` -> my_user
`http://instagram.com/user` -> user
`https://www.instagram.com/user` ->
`instagram.com/@user` -> user
`https://www.instagram.com/. /@user` -> error
`/@user` -> error
`instagram.com/user` -> user

The error message looks like this:

image

The message is translatable

image


This looks good to me!

**Pull request merged**